README.md
# duel-names
[![npm status](http://img.shields.io/npm/v/duel-names.svg)](https://www.npmjs.org/package/duel-names)
[![build status](https://secure.travis-ci.org/clux/duel-names.svg)](http://travis-ci.org/clux/duel-names)
[![dependency status](https://david-dm.org/clux/duel-names.svg)](https://david-dm.org/clux/duel-names)
[![coverage status](http://img.shields.io/coveralls/clux/duel-names.svg)](https://coveralls.io/r/clux/duel-names)

A mixin module for [Duel](https://www.npmjs.org/package/duel) that puts `roundName` function on the `Duel` class prototype that outputs the normal english names for duel rounds. This is here, rather than inside the duel repository for easy internationalization.

## Usage
Mixin with the Duel class:

```js
var Duel = require('duel');
var duelNames = require('duel-names');
Duel.attachNames(duelNames(require('duel-names/english')));

var d = new Duel(4, { last: Duel.WB });
d.roundName(d.matches[3].id);
'Bronze Final';
d.roundName(d.matches[2].id);
'Grand Final'
```

You can pass in a full match id, or a partial Id. The `id` must minimally contain `{ s: bracket, r: round }`.

## Localizing
Add a language file, consider submitting it to this repo (but with tests please), or just maintain the language file yourself.

## License
MIT-Licensed. See LICENSE file for details.


CHANGELOG.md
1.2.0 / 2015-11-30
==================
  * Added portugese and spanish tranlations - via @alexweber
  * Allow arbitrary length `single` and `doubleWinners` arrays for language compat - via @alexweber

1.1.2 / 2015-11-16
==================
  * Add `.npmignore`

1.1.1 / 2014-08-04
==================
  * Fix bug in norwegian language file for large tournaments

1.1.0 / 2014-08-04
==================
  * Changed structure so users can inject their own language file.=
  * Added norwegian language file to base bundle

1.0.0 / 2014-08-02
==================
  * first release - factored out of duel@0.5.0


test/language.js
var $ = require('interlude')
  , Duel = require('duel')
  , duelNames = require('..')
  , test = require('bandage');

module.exports = function testLanguage(language, np, opts) {
  var locfile = require('../' + language);

  test(language + ' ' + np, function *(mt) {
    mt.ok(locfile, 'have localization file');
    mt.ok(opts.bronzeFinal, 'have bronzeFinal');
    mt.ok(opts.rounds, 'have rounds');
    mt.eq(typeof opts.rounds, 'object','rounds is an obj');

    mt.ok(opts.winnerRounds, 'have winnerRounds');
    mt.eq(typeof opts.winnerRounds, 'object','winnerRounds is an obj');
    mt.ok(opts.loserRounds, 'have loserRounds');
    mt.eq(typeof opts.loserRounds, 'object','loserRounds is an obj');
    mt.equal(
      Object.keys(opts.loserRounds).length,
      Object.keys(opts.winnerRounds).length*2,
      'double the number of loserRounds passed in as winnerRounds'
    );

    Duel.attachNames(duelNames(locfile));
    mt.ok(Duel.prototype.roundName, 'attached roundName member');

    yield mt.test('single', function *(t) {
      var d = new Duel(np, { last: Duel.WB }) // using bronze final
      , gs = d.matches;

      var rnds = $.nub(gs.map($.get('id', 'r'))).sort($.compare());
      // prove we covered all the expected bases
      t.deepEqual(rnds,
        Object.keys(opts.rounds).map(Number),
        'number of rounds match passed in'
      );

      gs.forEach(function (m) {
        var r = m.id.r;
        var name = d.roundName(m.id);
        if (m.id.s === Duel.LB) {
          t.equal(name, opts.bronzeFinal, 'LB match has bronze final name');
        }
        else {
          t.equal(name, opts.rounds[r], 'Round ' + r + ' name is ' + opts.rounds[r]);
        }
      });
    });

    yield mt.test('double', function *(t) {
      var d = new Duel(np, { last: Duel.LB }) // using double final
        , gs = d.matches;
      // twice the amount of LB rounds as wb rounds
      var wbr = $.nub(d.findMatches({s: 1}).map($.get('id', 'r'))).sort($.compare());
      var lbr = $.nub(d.findMatches({s: 2}).map($.get('id', 'r'))).sort($.compare());

      t.equal(wbr.length*2, lbr.length, 'twice as many lb rounds');
      t.deepEqual(wbr,
        Object.keys(opts.winnerRounds).map(Number),
        'right number of wb rounds'
      );
      t.deepEqual(lbr,
        Object.keys(opts.loserRounds).map(Number),
        'right number of lb rounds'
      );

      gs.forEach(function (m) {
        var r = m.id.r;
        var name = d.roundName(m.id);
        if (m.id.s === Duel.LB) {
          t.equal(name, opts.loserRounds[r],
            'Round ' + r + ' LB name is ' + opts.loserRounds[r]
          );
        }
        else {
          t.equal(name, opts.winnerRounds[r],
            'Round ' + r + ' WB name is ' + opts.winnerRounds[r]
          );
        }
      });
    });
  });
};




test/spanish.test.js
var testLanguage = require('./language');

testLanguage('spanish', 32, {
  // Single elimination
  rounds: {
    1: 'Ronda de 32',
    2: 'Octavos de final',
    3: 'Cuartos de final',
    4: 'Semifinal',
    5: 'Gran final'
  },
  bronzeFinal: 'Partido del tercer lugar',
  // Double elimination names
  winnerRounds: {
    1: 'Ganadores Ronda de 32',
    2: 'Ganadores Octavos de final',
    3: 'Ganadores Cuartos de final',
    4: 'Ganadores Semifinal',
    5: 'Ganadores Final'
  },
  loserRounds: {
    1: 'Repesca Ronda de 16',
    2: 'Repesca Ronda Decisiva de 16',
    3: 'Repesca Ronda de 8',
    4: 'Repesca Ronda Decisiva de 8',
    5: 'Repesca Ronda de 4',
    6: 'Repesca Ronda Decisiva de 4',
    // special rounds
    7: 'Partido del cuarto lugar',
    8: 'Partido del tercer lugar',
    9: 'Final',
    10: 'Gran final'
  }
});


test/norwegian.test.js
var testLanguage = require('./language');

testLanguage('norwegian', 32, {
  // Single elimination
  rounds: {
    1: 'Sekstendelsfinaler',
    2: 'Åttendelsfinaler',
    3: 'Kvartfinaler',
    4: 'Semifinaler',
    5: 'Finale'
  },
  bronzeFinal: 'Bronsefinale',
  // Double elimination names
  winnerRounds: {
    1: 'Sekstendelsfinaler',
    2: 'Åttendelsfinaler',
    3: 'Kvartfinaler',
    4: 'Semifinaler',
    5: 'Finale'
  },
  loserRounds: {
    1: 'Tapernes åttendelsfinaler',
    2: 'Tapernes siste åttendelsfinaler',
    3: 'Tapernes kvartfinaler',
    4: 'Tapernes siste kvartfinaler',
    5: 'Tapernes semifinaler',
    6: 'Tapernes siste semifinaler',
    // special rounds
    7: 'Tapernes finale',
    8: 'Tapernes storfinale',
    9: 'Storfinale',
    10: 'Storfinale'
  }
});

testLanguage('norwegian', 512, {
  // Single elimination
  rounds: {
    1: '256-delsfinaler',
    2: 'Hundreogtjueåttedelsfinaler',
    3: 'Sekstifiredelsfinaler',
    4: 'Trettitodelsfinaler',
    5: 'Sekstendelsfinaler',
    6: 'Åttendelsfinaler',
    7: 'Kvartfinaler',
    8: 'Semifinaler',
    9: 'Finale'
  },
  bronzeFinal: 'Bronsefinale',
  // Double elimination names
  winnerRounds: {
    1: '256-delsfinaler',
    2: 'Hundreogtjueåttedelsfinaler',
    3: 'Sekstifiredelsfinaler',
    4: 'Trettitodelsfinaler',
    5: 'Sekstendelsfinaler',
    6: 'Åttendelsfinaler',
    7: 'Kvartfinaler',
    8: 'Semifinaler',
    9: 'Finale'
  },
  loserRounds: {
    1: 'Tapernes hundreogtjueåttedelsfinaler',
    2: 'Tapernes siste hundreogtjueåttedelsfinaler',
    3: 'Tapernes sekstifiredelsfinaler',
    4: 'Tapernes siste sekstifiredelsfinaler',
    5: 'Tapernes trettitodelsfinaler',
    6: 'Tapernes siste trettitodelsfinaler',
    7: 'Tapernes sekstendelsfinaler',
    8: 'Tapernes siste sekstendelsfinaler',
    9: 'Tapernes åttendelsfinaler',
    10: 'Tapernes siste åttendelsfinaler',
    11: 'Tapernes kvartfinaler',
    12: 'Tapernes siste kvartfinaler',
    13: 'Tapernes semifinaler',
    14: 'Tapernes siste semifinaler',
    // special rounds
    15: 'Tapernes finale',
    16: 'Tapernes storfinale',
    17: 'Storfinale',
    18: 'Storfinale'
  }
});


english.js
var appendRoundNum = function (str) {
  return function (num) {
    return str + " of " + num;
  };
};
var constant = function (str) {
  return function () {
    return str;
  };
};

exports.single = [
  // bronze final is a special case and must be the first entry
  constant("Bronze final"),
  // remaining entries are rounds in descending order of importance
  constant("Grand final"),
  constant("Semi-finals"),
  constant("Quarter-finals"),
  appendRoundNum("Round")
];

exports.doubleWinners = [
  constant("WB Final"),
  constant("WB Semi-finals"),
  constant("WB Quarter-finals"),
  appendRoundNum("WB Round")
];

exports.doubleLosers = [
  constant("Grand final"),          // Strong grand final (no prefix lest we spoil)
  constant("Grand final"),          // Potentially last game
  constant("LB Strong final"),      // 3rd place decider
  constant("LB Final"),             // 4th place decider
  appendRoundNum("LB Strong Round"),// last time there's X in LB
  appendRoundNum("LB Round")        // first time there's X in LB
];


spanish.js
var appendRoundNum = function (str) {
  return function (num) {
    return str + " de " + num;
  };
};
var constant = function (str) {
  return function () {
    return str;
  };
};

exports.single = [
  // bronze final is a special case and must be the first entry
  constant("Partido del tercer lugar"),
  // remaining entries are rounds in descending order of importance
  constant("Gran final"),
  constant("Semifinal"),
  constant("Cuartos de final"),
  constant("Octavos de final"),
  appendRoundNum("Ronda")
];

exports.doubleWinners = [
  constant("Ganadores Final"),
  constant("Ganadores Semifinal"),
  constant("Ganadores Cuartos de final"),
  constant("Ganadores Octavos de final"),
  appendRoundNum("Ganadores Ronda")
];

exports.doubleLosers = [
  constant("Gran final"),                     // Strong grand final (no prefix lest we spoil)
  constant("Final"),                          // Potentially last game
  constant("Partido del tercer lugar"),       // 3rd place decider
  constant("Partido del cuarto lugar"),       // 4th place decider
  appendRoundNum("Repesca Ronda Decisiva"),   // last time there's X in LB
  appendRoundNum("Repesca Ronda")             // first time there's X in LB
];


duel_names.js
module.exports = function (names) {
  var roundNameSingle = function (T, p, br, r) {
    if (br === T.LB) {
      return names.single[0]();
    }
    var last = names.single.length - 1;
    return names.single[(r + last > p) ? p - r + 1 : last](Math.pow(2, p - r + 1));
  };

  var roundNameDouble = function (T, p, br, r) {
    if (br === T.WB) {
      var last = names.doubleWinners.length - 1;
      return names.doubleWinners[(r + last > p) ? p - r : last](Math.pow(2, p - r + 1));
    }
    // gf rounds or lb final first, else treat like (strong?) round of X (idx 4 or 5)
    var lbIdx = (r >= 2*p - 3) ? 2*p - r : (4 + r%2);

    // round number 2n always has same number as 2n-1 because of feeding
    return names.doubleLosers[lbIdx](Math.pow(2, p -  Math.floor((r+1)/2)));
  };


  // can take a partial id where everything but the match number is left out
  // T is a constants object injected by Duel along with last bracket and Duel power.
  return function (T, last, p, partialId) {
    var br = partialId.s
      , r = partialId.r;

    // sanity
    if (!Number.isFinite(r) || r < 1 || [T.WB, T.LB].indexOf(br) < 0) {
      throw new Error("invalid partial id for roundName: " + partialId);
    }
    var invWB = (br === T.WB && r > p)
      , invSeLB = (last === T.WB && br >= T.LB && r !== 1)
      , invDeLB = (last === T.LB && r > 2*p);

    if (invWB || invSeLB || invDeLB) {
      var str = "br=" + br + ", last=" + last;
      throw new Error("invalid round " + r + " given for elimination: " + str);
    }

    return (last === T.WB ? roundNameSingle : roundNameDouble)(T, p, br, r);
  };
};


portuguese.js
var appendRoundNum = function (str) {
  return function (num) {
    return str + " de " + num;
  };
};
var constant = function (str) {
  return function () {
    return str;
  };
};

exports.single = [
  // bronze final is a special case and must be the first entry
  constant("Disputa do terceiro Lugar"),
  // remaining entries are rounds in descending order of importance
  constant("Grande final"),
  constant("Semifinal"),
  constant("Quartas de final"),
  appendRoundNum("Rodada")
];

exports.doubleWinners = [
  constant("Vencedores Final"),
  constant("Vencedores Semifinal"),
  constant("Vencedores Quartas de final"),
  appendRoundNum("Vencedores Rodada")
];

exports.doubleLosers = [
  constant("Grande final"),                                 // Strong grand final (no prefix lest we spoil)
  constant("Final"),                                        // Potentially last game
  constant("Repescagem Disputa do terceiro lugar"),         // 3rd place decider
  constant("Repescagem Disputa do quarto lugar"),           // 4th place decider
  appendRoundNum("Repescagem Rodada Decisiva"),             // last time there's X in LB
  appendRoundNum("Repescagem Rodada")                       // first time there's X in LB
];


norwegian.js
var specials = {
  '4': 'Semi',
  '8': 'Kvart',
  '16': 'Åttendels',
  '32': 'Sekstendels',
  '64': 'Trettitodels',
  '128': 'Sekstifiredels',
  '256': 'Hundreogtjueåttedels'
};

var ordinals = function (num) {
  return (num in specials) ? specials[num] : num/2 + '-dels';
};

var prependLosersThenOrdinal = function (losers, str) {
  return function (num) {
    return losers + ordinals(num).toLowerCase() + str;
  };
};

var prependOrdinal = function (str) {
  return function (num) {
    return ordinals(num) + str;
  };
};
var constant = function (str) {
  return function () {
    return str;
  };
};

exports.single = [
  constant('Bronsefinale'),
  constant('Finale'),
  prependOrdinal('finaler')
];

exports.doubleWinners = [
  constant('Finale'),
  prependOrdinal('finaler')
];

exports.doubleLosers = [
  constant('Storfinale'),
  constant('Storfinale'),
  constant('Tapernes storfinale'),
  constant('Tapernes finale'),
  prependLosersThenOrdinal('Tapernes siste ', 'finaler'),
  prependLosersThenOrdinal('Tapernes ', 'finaler')
];
